File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.java
import java.util.*;
class Solution {
    public int[] solution(String msg) {

        List<String> arr = new ArrayList<>();
        List<Integer> result = new ArrayList<>();
        
        for(int i=0;i<26;i++){
            arr.add(String.valueOf((char)('A' + i)));
        }
        
        for (int i=0;i<msg.length();i++){
            for (int j = arr.size() - 1;j>=0;j--){
                if (msg.substring(i).startsWith(arr.get(j))) {
                    i += arr.get(j).length() - 1;
                    result.add(j + 1);
                    if (i+1 < msg.length()) {
                        arr.add(arr.get(j) + msg.charAt(i + 1));
                    }
                    break;
                }
            }
        }
        
        int[] answer = new int[result.size()];
        
        for (int i=0;i<answer.length;i++){
            answer[i] = result.get(i);
        }
        
        
        
        
        return answer;
    }
}

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.cpp
#include <bits/stdc++.h>

using namespace std;

map<string,int> m;
string w,c;
int idx = 1;

vector<int> solution(string msg) {
    vector<int> answer;

    for (char i='A';i<='Z';i++) {
        m[string(1, i)] = idx++;
    }

    for (int i=0;i<msg.size();i++) {
        w = "";
        int j = i;

        while (j < msg.size() && m.count(msg.substr(i, j - i + 1))) j++;


        w = msg.substr(i, j - i);
        answer.push_back(m[w]);

        if (j < msg.size()) {
            c = msg.substr(i, j - i + 1);
            m[c] = idx++;
        }

        i = j-1;

    }

    return answer;
}